src/Discussion/Filter/AuthorFilterGambit.php

```diff
@@ -35,23 +31,39 @@ public function __construct(UserRepository $users)
         $this->users = $users;
     }
 
+    /**
+     * {@inheritdoc}
+     */
+    protected $pattern = 'author:(.+)';
+
     /**
      * {@inheritdoc}
      */
     protected function conditions(AbstractSearch $search, array $matches, $negate)
     {
-        if (! $search instanceof DiscussionSearch) {
-            throw new LogicException('This gambit can only be applied on a DiscussionSearch');
-        }
```

Comment:
I'm not sure it's that useful. Core's gambits aren't really used outside of core, so little risk of them being applied to the wrong search. And removing it here doesn't really affect external gambits, although with those, now that gambits are registered directly (and not via an event) it's much less likely to attach a gambit to the wrong searcher.

Reply:
Good point, doesn't hurt to have it though. Looking at the code, it seems to have been there because we can't explicitly typehint the conditions method with the exact search class, because the method signature has to be identical to the parent class's abstract method. And this bit of code also helps IDEs like phpStorm know the specific type of the $search object, which isn't really much useful at the moment considering these classes don't have anything more than their abstract parent (DiscussionSearch does have additional methods but none of them seem to be used anywhere).

Reply:
That's fair. In that case, should `WrappedFilter` be redesigned as `AbstractFilter`, so we can have a similar system for filtering? I'm not sure it'd be that useful, since you wouldn't really have "relevant posts" when filtering.

Reply:
While I'm all for consistency, I also don't really see it being of use.

Reply:
For that matter, do we have a use case for separate "search" classes for each model? As you noted, the specialized methods/fields on the DiscussionSearch one aren't used, and extensions can't add new methods onto it, so there's no use there. Furthermore, if we made AbstractSearch non-abstract and just used that, it wouldn't block us from switching back to  subclass of it eventually, since polymorphism.

Reply:
Indeed! That really seems to be the case, just need to make 100% sure those methods are not used, and then we could in a separate pull request look into  the `Query` namespace (just to avoid more drastic changes in this PR so that other devs can review more easily).

Reply:
Actually, never mind: `DiscussionSearch` has a `$defaultSort` field. And THAT could be applicable to filters too. Urgh....

I'll think some more about this, but I think we're nearing completion for this refactor.

Reply:
New proposal: https://github.com/flarum/core/issues/1355

Reply:
Done.